### PR TITLE
Westinm/downstreamapi-logging

### PR DESCRIPTION
**Pull request by westin-m**
Introduce logging to DonwstreamApi for errors that occur due to misconfigured appsettings (base url, relative path, typos in api) and for unauthenticated requests.

### CODE REVIEW

File: src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.cs [hunk 1]
```diff
@@ -23,21 +24,25 @@ internal partial class DownstreamApi : IDownstreamApi
         private readonly IOptionsMonitor<DownstreamApiOptions> _namedDownstreamApiOptions;
         private const string ScopesNotConfiguredInConfigurationOrViaDelegate = "IDW10107: Scopes need to be passed-in either by configuration or by the delegate overriding it. ";
         private const string Authorization = "Authorization";
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.cs:26]: If this is not used, let's remove it.

File: src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.Logger.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                    "[MsIdWeb] An error occurred during HTTP Request. " + 
+                    "ServiceName: {serviceName}, " +
+                    "BaseUrl: {BaseUrl}, " +
+                    "RelativePath: {RelativePath} ");
+
+            private static readonly Action<ILogger, Exception?> s_unauthenticatedApiCall =
+                LoggerMessage.Define(
+                    LogLevel.Information,
+                    DownstreamApiLoggingEventId.UnauthenticatedApiCall, 
+                    "[MsIdWeb] An unauthenticated call was made to the Api with null Scopes");
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.Logger.cs:30]: >Api [](http://example.com/codeflow?start=71&length=3)

nit: API

File: src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.Logger.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+
+            /// <summary>
+            /// Logger for handling options exceptions in DownstreamApi
+            /// </summary>
+            /// <param name="logger">ILogger</param>
+            /// <param name="ServiceName">Name of Api receiving request</param>
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.Logger.cs:37]: >Api [](http://example.com/codeflow?start=50&length=3)

nit: API

File: src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.Logger.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections;
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.Logger.cs:5]: nit: remove unused using

**Comment by westin-m** [comment]
> Looks good so far. Do you plan to do more or ready to move out of draft?

Ready to move out of draft

File: src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.HttpMethods.cs [hunk 5]
```diff
@@ -22,9 +22,23 @@ internal partial class DownstreamApi : IDownstreamApi
             where TOutput : class
         {
             DownstreamApiOptions effectiveOptions = MergeOptions(serviceName, downstreamApiOptionsOverride, HttpMethod.Get);
-            HttpResponseMessage response = await CallApiInternalAsync(serviceName, effectiveOptions, false, null, user, cancellationToken).ConfigureAwait(false);
-
-            return await DeserializeOutput<TOutput>(response, effectiveOptions).ConfigureAwait(false);
+            
+            try
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.HttpMethods.cs:26]: For the moment, you can merge

**Review by jmprieur — APPROVED** [review]
LGTM
Follow-up is to move the code in the .tt file.

**Pull request merged**